pkg/agent/controller/traceflow/packetin.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package traceflow
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/contiv/libOpenflow/openflow13"
+	"github.com/contiv/ofnet/ofctrl"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+
+	opsv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/ops/v1alpha1"
+)
+
+type JSONPatch struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
+}
+
+func (c *Controller) ReceivePacketIn(stopCh <-chan struct{}) {
+	ch := make(chan *ofctrl.PacketIn)
+	c.ofClient.SubscribePacketIn(1, ch)
+
+	wait.PollUntil(time.Second, func() (done bool, err error) {
+		pktIn := <-ch
+		oldTf, nodeResult, err := c.ParsePacketIn(pktIn)
+		if err != nil {
+			klog.Errorf("ParsePacketIn error: %+v", err)
+			return false, err
+		}
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			tf, err := c.traceflowClient.OpsV1alpha1().Traceflows().Get(oldTf.Name, v1.GetOptions{})
+			if err != nil {
+				klog.Warningf("Get traceflow failed: %+v", err)
+				return err
+			}
+			if oldTf.Status.Phase == opsv1alpha1.Succeeded {
+				tf.Status.Phase = oldTf.Status.Phase
+			}
+			tf.Status.Results = append(tf.Status.Results, *nodeResult)
+			tf, err = c.traceflowClient.OpsV1alpha1().Traceflows().Update(tf)
+			if err != nil {
+				klog.Warningf("Update traceflow failed: %+v", err)
+				return err
+			}
+			c.updateTagCache(tf)
+			klog.Infof("Updated traceflow %s: %+v", tf.Name, nodeResult)
+			return err
+		})
+		if err != nil {
+			klog.Errorf("Update traceflow error: %+v", err)
+			return false, err
+		}
+		return false, err
+	}, stopCh)
+}
+
+func (c *Controller) ParsePacketIn(pktIn *ofctrl.PacketIn) (*opsv1alpha1.Traceflow, *opsv1alpha1.NodeResult, error) {
+	matchers := pktIn.GetMatches()
+	var match *ofctrl.MatchField
+
+	// get cross node tag
+	if match = getMatchField(matchers, uint32(traceflowReg)); match == nil {
+		return nil, nil, errors.New("traceflow cross node tag not found")
+	}
+	rngTag := openflow13.NewNXRange(28, 31)
```

Comment:
use var ofTraceflowMarkRange = binding.Range{28, 31}

Reply:
I noticed you have defined a variable in pkg/agent/openflow/, and such variable should be used in both packetOut and packetIn, so it would be better to use the same one in different files to avoid misconfigurations.

Reply:
Can we export ofTraceflowMarkRange?
This means ofTraceflowMarkRange  -> OfTraceflowMarkRange

Reply:
Exported OfTraceflowMarkRange